depthai/plugin.cpp

```diff
@@ -0,0 +1,344 @@
+
+#include <cstdio>
+#include <iostream>
+
+#include "opencv2/opencv.hpp"
+
+// Inludes common necessary includes for development using depthai library
+#include "depthai/depthai.hpp"
+
+// ILLIXR includes
+#include "common/threadloop.hpp"
+#include "common/switchboard.hpp"
+#include "common/data_format.hpp"
+
+using namespace ILLIXR;
+
+class depthai : public plugin {
+public:
+    depthai(std::string name_, phonebook* pb_)
+        : plugin{name_, pb_}
+        , sb{pb->lookup_impl<switchboard>()}
+        , _m_imu_cam{sb->get_writer<imu_cam_type>("imu_cam")}
+        , _m_rgb_depth{sb->get_writer<rgb_depth_type>("rgb_depth")}
+        //Initialize DepthAI pipeline and device 
+        , p{createCameraPipeline()}
+        , d{p}
+        { 
+            d.startPipeline();
+            #ifndef NDEBUG
+                std::cout << "Depthai pipeline started" << std::endl;
+            #endif
+            color = d.getOutputQueue("preview", 1, false);
+            depthQueue = d.getOutputQueue("depth", 1, false);
+            rectifLeftQueue = d.getOutputQueue("rectified_left", 1, false);
+            rectifRightQueue = d.getOutputQueue("rectified_right", 1, false);
+            imuQueue = d.getOutputQueue("imu", 1, false);
+            std::function<void(void)> imu_callback = [&](){callback();};
+            imuQueue->addCallback(imu_callback);
+        }
+
+    void callback(){  
+        std::lock_guard<std::mutex> lock(mutex);
```

Comment:
Is the `callback` going to be called from multiple threads? If not, we should remove this lock and mutex.

Reply:
It shouldn't be, but I need to verify. Also, I wanted to investigate this comment regarding that
https://github.com/ILLIXR/ILLIXR/blob/287f6c698b2a479545d60f409cb2fcfa70a96f2f/realsense/plugin.cpp#L37

> // This lock guarantees that concurrent invocations of `callback` are serialized.
            // Even if the API does not invoke `callback` in parallel, this is still important for the memory-model.
            // Without this lock, prior invocations of `callback` are not necessarily "happens-before" ordered, so accessing persistent variables constitutes a data-race, which is undefined behavior in the C++ memory model.